// @primer/react/lib/Token/_RemoveTokenButton.js
'use strict';

var octiconsReact = require('@primer/octicons-react');
var clsx = require('clsx');
var TokenBase = require('./TokenBase.js');
var _RemoveTokenButton_module = require('./_RemoveTokenButton.module.css.js');
var BoxWithFallback = require('../internal/components/BoxWithFallback.js');
var jsxRuntime = require('react/jsx-runtime');

const getTokenButtonIconSize = size => parseInt(TokenBase.tokenSizes[size || TokenBase.defaultTokenSize], 10) * 0.75;
const RemoveTokenButton = ({
  'aria-label': ariaLabel,
  isParentInteractive,
  size = TokenBase.defaultTokenSize,
  className,
  borderOffset = 0,
  as: _as,
  ...rest
}) => {
  // eslint-disable-next-line react-compiler/react-compiler
  delete rest.children;
  return /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
    as: isParentInteractive ? 'span' : 'button',
    tabIndex: isParentInteractive ? -1 : undefined,
    "aria-label": !isParentInteractive ? 'Remove token' : ariaLabel,
    "data-size": size,
    className: clsx.clsx(_RemoveTokenButton_module.TokenButton, className),
    style: {
      transform: `translate(${borderOffset}px, -${borderOffset}px)`
    },
    ...rest,
    children: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.XIcon, {
      size: getTokenButtonIconSize(size)
    })
  });
};
RemoveTokenButton.displayName = "RemoveTokenButton";

module.exports = RemoveTokenButton;
