// @primer/react/lib/RelativeTime/RelativeTime.js
'use strict';

require('../node_modules/@github/relative-time-element/dist/relative-time-element-define.js');
var createComponent = require('../utils/create-component.js');
var jsxRuntime = require('react/jsx-runtime');
var relativeTimeElement = require('../node_modules/@github/relative-time-element/dist/relative-time-element.js');

const RelativeTimeComponent = createComponent.createComponent(relativeTimeElement.RelativeTimeElement, 'relative-time');
const localeOptions = {
  month: 'short',
  day: 'numeric',
  year: 'numeric'
};
function RelativeTime({
  date,
  datetime,
  children,
  noTitle,
  ...props
}) {
  var _date;
  if (datetime) date = new Date(datetime);
  return /*#__PURE__*/jsxRuntime.jsx(RelativeTimeComponent, {
    ...props,
    date: date,
    "no-title": noTitle ? '' : undefined,
    children: children || ((_date = date) === null || _date === void 0 ? void 0 : _date.toLocaleDateString('en', localeOptions)) || ''
  });
}
RelativeTime.displayName = "RelativeTime";

module.exports = RelativeTime;
