// @primer/react/lib/internal/components/CheckboxOrRadioGroup/CheckboxOrRadioGroupLabel.js
'use strict';

var React = require('react');
var _VisuallyHidden = require('../../../_VisuallyHidden.js');
var CheckboxOrRadioGroupContext = require('./CheckboxOrRadioGroupContext.js');
var CheckboxOrRadioGroup_module = require('./CheckboxOrRadioGroup.module.css.js');
var index = require('../../../Stack/index.js');
var clsx = require('clsx');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const CheckboxOrRadioGroupLabel = ({
  children,
  className,
  visuallyHidden = false,
  sx
}) => {
  const {
    required,
    disabled
  } = React__default.default.useContext(CheckboxOrRadioGroupContext);
  return /*#__PURE__*/jsxRuntime.jsx(_VisuallyHidden, {
    className: clsx.clsx(className, CheckboxOrRadioGroup_module.RadioGroupLabel),
    isVisible: !visuallyHidden,
    title: required ? 'required field' : undefined,
    "data-label-disabled": disabled ? '' : undefined,
    sx: sx,
    children: required ? /*#__PURE__*/jsxRuntime.jsxs(index.Stack, {
      direction: "horizontal",
      gap: "none",
      children: [/*#__PURE__*/jsxRuntime.jsx("div", {
        className: CheckboxOrRadioGroup_module.GroupLabelChildren,
        children: children
      }), /*#__PURE__*/jsxRuntime.jsx("span", {
        children: "*"
      })]
    }) : children
  });
};
CheckboxOrRadioGroupLabel.displayName = "CheckboxOrRadioGroupLabel";

module.exports = CheckboxOrRadioGroupLabel;
