// @primer/react/lib/Stack/Stack.js
'use strict';

var React = require('react');
var getResponsiveAttributes = require('../internal/utils/getResponsiveAttributes.js');
var Stack_module = require('./Stack.module.css.js');
var clsx = require('clsx');
var BoxWithFallback = require('../internal/components/BoxWithFallback.js');
var jsxRuntime = require('react/jsx-runtime');

const Stack = /*#__PURE__*/React.forwardRef(({
  as,
  children,
  align = 'stretch',
  direction = 'vertical',
  gap,
  justify = 'start',
  padding = 'none',
  wrap = 'nowrap',
  className,
  ...rest
}, forwardedRef) => {
  return /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
    as: as,
    ref: forwardedRef,
    ...rest,
    className: clsx.clsx(className, Stack_module.Stack),
    ...getResponsiveAttributes.getResponsiveAttributes('gap', gap),
    ...getResponsiveAttributes.getResponsiveAttributes('direction', direction),
    ...getResponsiveAttributes.getResponsiveAttributes('align', align),
    ...getResponsiveAttributes.getResponsiveAttributes('wrap', wrap),
    ...getResponsiveAttributes.getResponsiveAttributes('justify', justify),
    ...getResponsiveAttributes.getResponsiveAttributes('padding', padding),
    children: children
  });
});
const StackItem = /*#__PURE__*/React.forwardRef(({
  as,
  children,
  grow,
  className,
  ...rest
}, forwardedRef) => {
  return /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
    as: as,
    ref: forwardedRef,
    ...rest,
    className: clsx.clsx(className, Stack_module.StackItem),
    ...getResponsiveAttributes.getResponsiveAttributes('grow', grow),
    children: children
  });
});

exports.Stack = Stack;
exports.StackItem = StackItem;
