// @primer/react/lib/ConfirmationDialog/ConfirmationDialog.js
'use strict';

var React = require('react');
var client = require('react-dom/client');
var ThemeProvider = require('../ThemeProvider.js');
var behaviors = require('@primer/behaviors');
var Dialog = require('../Dialog/Dialog.js');
var useFocusZone = require('../hooks/useFocusZone.js');
var BaseStyles = require('../BaseStyles.js');
var ConfirmationDialog_module = require('./ConfirmationDialog.module.css.js');
var jsxRuntime = require('react/jsx-runtime');

const ConfirmationHeader = ({
  title,
  onClose,
  dialogLabelId
}) => {
  const onCloseClick = React.useCallback(() => {
    onClose('close-button');
  }, [onClose]);
  return /*#__PURE__*/jsxRuntime.jsxs("div", {
    className: ConfirmationDialog_module.ConfirmationHeader,
    children: [/*#__PURE__*/jsxRuntime.jsx("h1", {
      id: dialogLabelId,
      children: title
    }), /*#__PURE__*/jsxRuntime.jsx(Dialog.Dialog.CloseButton, {
      onClose: onCloseClick
    })]
  });
};
ConfirmationHeader.displayName = "ConfirmationHeader";
const ConfirmationBody = ({
  children
}) => {
  return /*#__PURE__*/jsxRuntime.jsx("div", {
    className: ConfirmationDialog_module.ConfirmationBody,
    children: children
  });
};
ConfirmationBody.displayName = "ConfirmationBody";
const ConfirmationFooter = ({
  footerButtons
}) => {
  const {
    containerRef: footerRef
  } = useFocusZone.useFocusZone({
    bindKeys: behaviors.FocusKeys.ArrowHorizontal | behaviors.FocusKeys.Tab,
    focusInStrategy: 'closest'
  });

  // Must have exactly 2 buttons!
  return /*#__PURE__*/jsxRuntime.jsx("div", {
    ref: footerRef,
    className: ConfirmationDialog_module.ConfirmationFooter,
    children: /*#__PURE__*/jsxRuntime.jsx(Dialog.Dialog.Buttons, {
      buttons: footerButtons !== null && footerButtons !== void 0 ? footerButtons : []
    })
  });
};
ConfirmationFooter.displayName = "ConfirmationFooter";
/**
 * A ConfirmationDialog is a special kind of dialog with more rigid behavior. It
 * is used to confirm a user action. ConfirmationDialogs always have exactly
 * two buttons: one to cancel the action and one to confirm it. No custom
 * rendering capabilities are provided for ConfirmationDialog.
 */
const ConfirmationDialog = props => {
  const {
    onClose,
    title,
    cancelButtonContent = 'Cancel',
    confirmButtonContent = 'OK',
    confirmButtonType = 'normal',
    children,
    className,
    width = 'medium',
    height
  } = props;
  const onCancelButtonClick = React.useCallback(() => {
    onClose('cancel');
  }, [onClose]);
  const onConfirmButtonClick = React.useCallback(() => {
    onClose('confirm');
  }, [onClose]);
  const isConfirmationDangerous = confirmButtonType === 'danger';
  const cancelButton = {
    content: cancelButtonContent,
    onClick: onCancelButtonClick,
    autoFocus: isConfirmationDangerous
  };
  const confirmButton = {
    content: confirmButtonContent,
    buttonType: confirmButtonType,
    onClick: onConfirmButtonClick,
    autoFocus: !isConfirmationDangerous
  };
  const footerButtons = [cancelButton, confirmButton];
  return /*#__PURE__*/jsxRuntime.jsx(Dialog.Dialog, {
    onClose: onClose,
    title: title,
    footerButtons: footerButtons,
    role: "alertdialog",
    width: width,
    height: height,
    className: className,
    renderHeader: ConfirmationHeader,
    renderBody: ConfirmationBody,
    renderFooter: ConfirmationFooter,
    children: children
  });
};
ConfirmationDialog.displayName = "ConfirmationDialog";
let hostElement = null;
async function confirm(themeProps, options) {
  const {
    content,
    ...confirmationDialogProps
  } = options;
  return new Promise(resolve => {
    hostElement ||= document.createElement('div');
    if (!hostElement.isConnected) document.body.append(hostElement);
    const root = client.createRoot(hostElement);
    const onClose = gesture => {
      root.unmount();
      if (gesture === 'confirm') {
        resolve(true);
      } else {
        resolve(false);
      }
    };
    root.render(/*#__PURE__*/jsxRuntime.jsx(ThemeProvider.ThemeProvider, {
      ...themeProps,
      children: /*#__PURE__*/jsxRuntime.jsx(BaseStyles, {
        children: /*#__PURE__*/jsxRuntime.jsx(ConfirmationDialog, {
          ...confirmationDialogProps,
          onClose: onClose,
          children: content
        })
      })
    }));
  });
}

/**
 * This hook takes no parameters and returns an `async` function, `confirm`. When `confirm`
 * is called, it shows the confirmation dialog. When the dialog is dismissed, a promise is
 * resolved with `true` or `false` depending on whether or not the confirm button was used.
 */
function useConfirm() {
  const {
    theme,
    colorMode,
    dayScheme,
    nightScheme
  } = ThemeProvider.useTheme();
  const result = React.useCallback(options => {
    const themeProps = {
      theme,
      colorMode,
      dayScheme,
      nightScheme
    };
    return confirm(themeProps, options);
  }, [theme, colorMode, dayScheme, nightScheme]);
  return result;
}

exports.ConfirmationDialog = ConfirmationDialog;
exports.useConfirm = useConfirm;
