// @primer/behaviors/dist/cjs/index.js
'use strict';

var anchoredPosition = require('./anchored-position.js');
var focusTrap = require('./focus-trap.js');
var focusZone = require('./focus-zone.js');
var scrollIntoView = require('./scroll-into-view.js');
var dimensions = require('./dimensions.js');



exports.getAnchoredPosition = anchoredPosition.getAnchoredPosition;
exports.focusTrap = focusTrap.focusTrap;
Object.defineProperty(exports, "FocusKeys", {
	enumerable: true,
	get: function () { return focusZone.FocusKeys; }
});
exports.activeDescendantActivatedDirectly = focusZone.activeDescendantActivatedDirectly;
exports.activeDescendantActivatedIndirectly = focusZone.activeDescendantActivatedIndirectly;
exports.focusZone = focusZone.focusZone;
exports.hasActiveDescendantAttribute = focusZone.hasActiveDescendantAttribute;
exports.isActiveDescendantAttribute = focusZone.isActiveDescendantAttribute;
exports.scrollIntoView = scrollIntoView.scrollIntoView;
exports.offset = dimensions.offset;
exports.overflowOffset = dimensions.overflowOffset;
exports.overflowParent = dimensions.overflowParent;
exports.positionedOffset = dimensions.positionedOffset;
