// @primer/react/lib/SplitPageLayout/SplitPageLayout.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var PageLayout = require('../PageLayout/PageLayout.js');

const Root = props => {
  return /*#__PURE__*/jsxRuntime.jsx(PageLayout.PageLayout, {
    containerWidth: "full",
    padding: "none",
    columnGap: "none",
    rowGap: "none",
    _slotsConfig: {
      header: Header,
      footer: Footer
    },
    ...props
  });
};
Root.displayName = "Root";
Root.displayName = 'SplitPageLayout';

// ----------------------------------------------------------------------------
// SplitPageLayout.Header

const Header = ({
  padding = 'normal',
  divider = 'line',
  ...props
}) => {
  // eslint-disable-next-line primer-react/direct-slot-children
  return /*#__PURE__*/jsxRuntime.jsx(PageLayout.PageLayout.Header, {
    padding: padding,
    divider: divider,
    ...props
  });
};
Header.displayName = "Header";
Header.displayName = 'SplitPageLayout.Header';

// ----------------------------------------------------------------------------
// SplitPageLayout.Content

const Content = ({
  width = 'large',
  padding = 'normal',
  ...props
}) => {
  return /*#__PURE__*/jsxRuntime.jsx(PageLayout.PageLayout.Content, {
    width: width,
    padding: padding,
    ...props
  });
};
Content.displayName = "Content";
Content.displayName = 'SplitPageLayout.Content';

// ----------------------------------------------------------------------------
// SplitPageLayout.Pane

const Pane = ({
  position = 'start',
  sticky = true,
  padding = 'normal',
  divider = 'line',
  ...props
}) => {
  return /*#__PURE__*/jsxRuntime.jsx(PageLayout.PageLayout.Pane, {
    position: position,
    sticky: sticky,
    padding: padding,
    divider: divider,
    ...props
  });
};
Pane.displayName = "Pane";
Pane.displayName = 'SplitPageLayout.Pane';

// ----------------------------------------------------------------------------
// SplitPageLayout.Footer

const Footer = ({
  padding = 'normal',
  divider = 'line',
  ...props
}) => {
  // eslint-disable-next-line primer-react/direct-slot-children
  return /*#__PURE__*/jsxRuntime.jsx(PageLayout.PageLayout.Footer, {
    padding: padding,
    divider: divider,
    ...props
  });
};
Footer.displayName = "Footer";
Footer.displayName = 'SplitPageLayout.Footer';

// ----------------------------------------------------------------------------
// Export

const SplitPageLayout = Object.assign(Root, {
  Header,
  Content,
  Pane,
  Footer
});

exports.Content = Content;
exports.Footer = Footer;
exports.Header = Header;
exports.Pane = Pane;
exports.Root = Root;
exports.SplitPageLayout = SplitPageLayout;
